src/provider/cloud-element-templates/properties/CustomProperties.js

```diff
@@ -72,18 +80,64 @@ export function CustomProperties(props) {
   const groups = [];
 
   const {
-    id
+    id,
+    properties,
+    groups: propertyGroups
   } = elementTemplate;
 
+  // (1) group properties by group id
```

Comment:
Playing around with this, I realized that there is no explicit way to define the ordering (e.g., which section would appear first / which one last) of the sections. I could imagine that this is important to properly define the UX of let's say Connector (e.g., first define the URL of a REST service before adding Header fields seems natural).

Was this discussed?

Reply:
Same applies to the `C7 Template Groups` PR of course

Reply:
No, this was not discussed in particular.

The ordering is defined by the order of the `properties`, at least that is what I would expect, I didn't create a test for it, but the [`groupBy` helper](https://github.com/bpmn-io/bpmn-js-properties-panel/pull/564/files#diff-3c1092b38e059ceff7ce76978ca0e37e08337db1acb59aeff93b8b4d6cba1f16R683) should respect that order.

So, if the first property belongs to group 1, group 1 will be shown first, and so on. The properties determine the order, as it was beforehand.

I think my intention here was to keep things simple, without coming up with complex ordering functionality inside the `groups`. If we invent something like that, we would have to make sure that the ordering is always side effect-free, something that can cause some complexity (e.g. cf. the way how [electron solves menu items ordering](https://www.electronjs.org/docs/v14-x-y/api/menu#menu-item-position))

Reply:
I also have the feeling, whatever we come up with to explicitly order template groups, that should work kind of equally (syntax-wise) to order my general properties panel groups, which we don't have yet (cf. https://github.com/bpmn-io/bpmn-js-properties-panel/issues/523).

Maybe a good follow-up task, since I believe we should handle this in an orderly manner. But I also have the sense that this would be a nice usability improvement for template creators. Maybe something we could ask for feedback (especially in the Connectors department). 👍

Reply:
Not sure if I am missing the point here but the order of groups should be mandated by the following:

* Order of apperance in the template (first defined, first rendered)
* Overruled by [group definitions](https://github.com/bpmn-io/bpmn-js-properties-panel/pull/564/files#diff-f4cf7fe9d4214e16a89074330306e719da28548aec922eefaaa4e056cb08cf92R353)

Same applies for properties defined; => order of definition = order of apperance.

Reply:
Ups - I missed that the ordering can be defined by the order in the template :-) Then - all good!

A test case for that would be superb IMO.

Reply:
Thanks for the clarification 👍  I added a test case to verify

> Order of appearance in the template (first defined, first rendered)

If we would go for the second rule @nikku mentioned (`Overruled by group definitions`), that would _always_ apply since the `group` needs to be defined in any case to be displayed.